fastvideo/v1/pipelines/stages/base.py

```diff
@@ -0,0 +1,125 @@
+from fastvideo.v1.inference_args import InferenceArgs
+from fastvideo.v1.logger import init_logger
+
+logger = init_logger(__name__)
+
+
+class PipelineStage(ABC):
+    """
+    Abstract base class for all pipeline stages.
+    
+    A pipeline stage represents a discrete step in the diffusion process that can be
+    composed with other stages to create a complete pipeline. Each stage is responsible
+    for a specific part of the process, such as prompt encoding, latent preparation, etc.
+    """
+    
+    def __init__(self, enable_logging: bool = False):
+        """
+        Initialize the pipeline stage.
+        
+        Args:
+            enable_logging: Whether to enable logging for this stage.
+        """
+        self._enable_logging = enable_logging
+        self._stage_name = self.__class__.__name__
+        self._logger = init_logger(f"fastvideo.v1.pipelines.stages.{self._stage_name}")
+    
+    @property
+    def device(self) -> torch.device:
+        """Get the device for this stage."""
+        return torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    
+    def set_logging(self, enable: bool):
+        """
+        Enable or disable logging for this stage.
+        
+        Args:
+            enable: Whether to enable logging.
+        """
+        self._enable_logging = enable
+    
+    def __call__(
+        self,
+        batch: ForwardBatch,
+        inference_args: InferenceArgs,
+    ) -> ForwardBatch:
+        """
+        Execute the stage's processing on the batch with optional logging.
+        Should not be overridden by subclasses.
+        
+        Args:
+            batch: The current batch information.
+            inference_args: The inference arguments.
+            
+        Returns:
+            The updated batch information after this stage's processing.
+        """
+        if self._enable_logging:
+            self._logger.info(f"[{self._stage_name}] Starting execution")
+            start_time = time.time()
+            
+            try:
+                # Call the actual implementation
+                result = self._call_implementation(batch, inference_args)
+                
+                execution_time = time.time() - start_time
+                self._logger.info(f"[{self._stage_name}] Execution completed in {execution_time * 1000:.2f} ms")
+                
+                return result
+            except Exception as e:
+                execution_time = time.time() - start_time
+                self._logger.error(f"[{self._stage_name}] Error during execution after {execution_time * 1000:.2f} ms: {e}")
+                self._logger.error(f"[{self._stage_name}] Traceback: {traceback.format_exc()}")
+                
+                # Re-raise the exception
+                raise
+        else:
+            # Just call the implementation directly if logging is disabled
+            return self._call_implementation(batch, inference_args)
```

Comment:
For better user experience it's better to validate input and output batch (e.g., check expected keys) before and after calling the implementation.